server/src/main/java/org/opensearch/index/query/BaseQueryRewriteContext.java

```diff
@@ -0,0 +1,135 @@
+        NamedXContentRegistry xContentRegistry,
+        NamedWriteableRegistry writeableRegistry,
+        Client client,
+        LongSupplier nowInMillis,
+        boolean validate
+    ) {
+
+        this.xContentRegistry = xContentRegistry;
+        this.writeableRegistry = writeableRegistry;
+        this.client = client;
+        this.nowInMillis = nowInMillis;
+        this.validate = validate;
+    }
+
+    /**
+     * The registry used to build new {@link XContentParser}s. Contains registered named parsers needed to parse the query.
+     */
+    public NamedXContentRegistry getXContentRegistry() {
+        return xContentRegistry;
+    }
+
+    /**
+     * Returns the time in milliseconds that is shared across all resources involved. Even across shards and nodes.
+     */
+    public long nowInMillis() {
+        return nowInMillis.getAsLong();
+    }
+
+    public NamedWriteableRegistry getWriteableRegistry() {
+        return writeableRegistry;
+    }
+
+    /**
+     * Returns an instance of {@link QueryShardContext} if available of null otherwise
+     */
+    public QueryShardContext convertToShardContext() {
+        return null;
+    }
+
+    /**
+     * Registers an async action that must be executed before the next rewrite round in order to make progress.
+     * This should be used if a rewriteabel needs to fetch some external resources in order to be executed ie. a document
+     * from an index.
+     */
+    public void registerAsyncAction(BiConsumer<Client, ActionListener<?>> asyncAction) {
+        asyncActions.add(asyncAction);
+    }
+
+    /**
+     * Returns <code>true</code> if there are any registered async actions.
+     */
+    public boolean hasAsyncActions() {
+        return asyncActions.isEmpty() == false;
+    }
+
+    /**
+     * Executes all registered async actions and notifies the listener once it's done. The value that is passed to the listener is always
+     * <code>null</code>. The list of registered actions is cleared once this method returns.
+     */
+    public void executeAsyncActions(ActionListener listener) {
+        if (asyncActions.isEmpty()) {
+            listener.onResponse(null);
+        } else {
+            CountDown countDown = new CountDown(asyncActions.size());
+            ActionListener<?> internalListener = new ActionListener() {
+                @Override
+                public void onResponse(Object o) {
+                    if (countDown.countDown()) {
+                        listener.onResponse(null);
+                    }
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    if (countDown.fastForward()) {
+                        listener.onFailure(e);
+                    }
+                }
+            };
+            // make a copy to prevent concurrent modification exception
+            List<BiConsumer<Client, ActionListener<?>>> biConsumers = new ArrayList<>(asyncActions);
```

Comment:
Is ConcurrentModificationException exception while iterating over the collection is the only concern here? Depending on what you're trying to avoid you may consider more robust approaches, like use CopyOnWriteArrayList instead of ArrayList.

Reply:
every time this method is called, it will create new ArrayList, so this thread is safe. Even when this method called multiple times, it will create multiple ArrayList. It shouldn't have multiple writes on the same position.

Reply:
> Even when this method called multiple times, it will create multiple ArrayList. It shouldn't have multiple writes on the same position.

The issue is not this method being called multiple times, it's _other_ methods that can modify the `asyncActions` List during the same time this method is executing.  In particular, if someone (on a different thread) calls `registerAsyncAction()` to add a new action in between when you copy it and clear it, you might delete the newly registered action by the other thread.

The current code is not thread safe. I realize you didn't write it, but since you're touching it you should fix it :)